### PR TITLE
uplift `clippy::clone_double_ref` to rustc

### LINKED ISSUE #109451: Uplift `clippy::clone_double_ref` to `rustc`
In #109429, a snippet of code involves cloning references outputted a confusing compiler error:

```rust
use std::collections::hash_map::DefaultHasher;
use std::collections::HashMap;
use std::hash::BuildHasher;
use std::hash::Hash;

pub struct Hash128_1;

impl BuildHasher for Hash128_1 {
    type Hasher = DefaultHasher;
    fn build_hasher(&self) -> DefaultHasher { DefaultHasher::default() }
}

#[allow(unused)]
pub fn hashmap_copy<T, U>(
    map: &HashMap<T, U, Hash128_1>,
) where T: Hash + Clone, U: Clone
{
    let mut copy: Vec<U> = map.clone().into_values().collect();
}

pub fn make_map() -> HashMap<String, i64, Hash128_1>
{
    HashMap::with_hasher(Hash128_1)
}
```

```
error[E0507]: cannot move out of a shared reference
  --> src/lib.rs:19:28
   |
19 |     let mut copy: Vec<U> = map.clone().into_values().collect();
   |                            ^^^^^^^^^^^^-------------
   |                            |           |
   |                            |           value moved due to this method call
   |                            move occurs because value has type `HashMap<T, U, Hash128_1>`, which does not implement the `Copy` trait
   |
note: `HashMap::<K, V, S>::into_values` takes ownership of the receiver `self`, which moves value
```

Changing the diagnostic to provide help when we are trying to use a function that takes `self` after a `.clone()` call is nontrivial, since this is MIR borrowck and to my knowledge it is hard to know if it comes from the return value of a `.clone()` call.

Ideally we should have an error/warning like the one from clippy:

```
error: using `clone` on a double-reference; this will copy the reference of type `&HashMap<T, U, Hash128_1>` instead of cloning the inner type
  --> src/lib.rs:19:28
   |
19 |     let mut copy: Vec<U> = map.clone().into_values().collect();
   |                            ^^^^^^^^^^^
```

I think there are other ways this can be confusing, so uplifting can help with many issues related to this and not just the example I linked.

----

cc @rust-lang/lang: would uplifting be a good idea? I'd be willing to work on this if this gets approval.

@rustbot label +A-lint +T-compiler +T-lang +D-newcomer-roadblock

### PR DESCRIPTION
**Pull request by fee1-dead**
This closes #109451, but is only part 1 for fixing #109429. The second part would be to run a selection of lints even after borrowck had errors.

https://github.com/rust-lang/rust/blob/b9535c0b7d64290b27c4a116262402cd12f77833/compiler/rustc_interface/src/passes.rs#L815-L822

Note: both suggestions are pretty bad.. Why don't we just suggest dereferencing?

---

For Clippy:

changelog: Moves: Uplifted `clippy::clone_double_ref` into rustc

### CODE REVIEW

**Comment by fee1-dead** [comment]
Turns out that `noop_method_call` is a thing already. This should probably be merged to that.

File: src/tools/clippy/tests/ui/unnecessary_clone.stderr [hunk 1]
```diff
@@ -75,13 +81,13 @@ LL |         let _ = &mut encoded.clone();
...
-   |                      ~~~~~~~~~~~~~~~~~~~
+   |                      +++       ~~~~~~~~~
 help: or try being explicit if you are sure, that you want to clone a reference
    |
 LL |         let _ = &mut <&[u8]>::clone(encoded);
-   |                      ~~~~~~~~~~~~~~~~~~~~~~~
+   |                      +++++++++++++++       ~
 
-error: using `clone` on a double-reference; this will copy the reference of type `&[u8]` instead of cloning the inner type
+error: using `clone` on a double-reference, which copies the reference of type `[u8]` instead of cloning the inner type
```
Comment (flip1995) [inline src/tools/clippy/tests/ui/unnecessary_clone.stderr:90]: This seems to still test the `clone_double_ref` lint inside the Clippy UI tests. All tests that are there explicitly for this lint should be moved to rustc UI tests.

**Review by flip1995 — COMMENTED** [review]
Clippy specific review.

File: compiler/rustc_lint/src/noop_method_call.rs [hunk 2]
```diff
@@ -31,18 +31,48 @@ declare_lint! {
     /// calling `clone` on a `&T` where `T` does not implement clone, actually doesn't do anything
     /// as references are copy. This lint detects these calls and warns the user about them.
     pub NOOP_METHOD_CALL,
-    Allow,
+    Warn,
```
Comment (compiler-errors) [inline compiler/rustc_lint/src/noop_method_call.rs:34]: I feel like this change and the associated changes to lib/compiler should be done in a separate PR.

Or at least the PR title and description need to be updated to reflect that this is also bumping NOOP_METHOD_CALL to warn.

File: compiler/rustc_lint/src/noop_method_call.rs [hunk 3]
```diff
@@ -31,18 +31,48 @@ declare_lint! {
...
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Since `Foo` doesn't implement `Clone`, running `.clone()` only dereferences the double
+    /// reference, instead of cloning the inner type which should be what was intended.
+    pub SUSPICIOUS_DOUBLE_REF_OP,
+    Warn,
+    "using `clone` on `&&T`"
```
Comment (compiler-errors) [inline compiler/rustc_lint/src/noop_method_call.rs:60]: This lint isn't `Clone` specific as implemented below?

File: tests/ui/underscore-imports/cycle.rs [hunk 4]
```diff
@@ -1,6 +1,7 @@
 // Check that cyclic glob imports are allowed with underscore imports
 
 // check-pass
+#![allow(noop_method_call)]
```
Comment (compiler-errors) [inline tests/ui/underscore-imports/cycle.rs:4]: I would actually prefer if we added the `WARN`s to this UI test and others.

File: tests/ui/lint/noop-method-call.stderr [hunk 5]
```diff
@@ -8,25 +8,49 @@ LL |     let non_clone_type_ref_clone: &PlainType<u32> = non_clone_type_ref.clon
...
 warning: call to `.deref()` on a reference in this situation does nothing
-  --> $DIR/noop-method-call.rs:28:63
+  --> $DIR/noop-method-call.rs:27:63
    |
 LL |     let non_deref_type_deref: &PlainType<u32> = non_deref_type.deref();
    |                                                               ^^^^^^^^ unnecessary method call
    |
    = note: the type `&PlainType<u32>` which `deref` is being called on is the same as the type returned from `deref`, so the method call does not do anything and can be removed
 
+warning: using `.deref()` on a double reference, which copies `&PlainType<u32>` instead of dereferencing the inner type
```
Comment (compiler-errors) [inline tests/ui/lint/noop-method-call.stderr:34]:
```suggestion
warning: using `.deref()` on a double reference, which returns `&PlainType<u32>` instead of dereferencing the inner type
```
